File: src/lib/db/sessions.ts
// Typed wrappers over the Rust `sessions_*` commands (local SQLite session
// history). Two near-identical shapes on purpose: `SessionRow` (camelCase) is
// the JS→Rust INSERT input — Tauri's invoke layer expects camelCase keys and
// matches them to the command's snake_case parameters — while `SessionRecord`
// (snake_case) is serde's response shape on reads. Don't merge them.

import { invoke } from '@tauri-apps/api/core'

export type SessionRow = {
  id: string
  startedAt: number
  endedAt: number
  totalMinutes: number | null
  // Exact accumulated awake study duration. New writers keep sub-minute
  // residue here while totalMinutes remains the derived display/stats count.
  totalDurationMs: number | null
  // JSON-array string of every ed_pubkey_hex observed via signed-hello in
  // this session, sorted lexicographically (canonical regardless of join
  // order). NULL when no hello was received — solo session or pre-V1-P9.
  peerPubkeys: string | null
  // Canonical JSON object of ed_pubkey_hex -> cumulative authenticated
  // overlap milliseconds. New rows always write an object (including `{}`);
  // null means legacy/unknown precision.
  peerPresenceMs?: string | null
  // V2-P8 report fields. Populated by the leave handler once the post-
  // session report runs; the V1 lifecycle insert leaves them null and the
  // Rust upsert preserves null-overrides via COALESCE.
  declaredTopic?: string | null
  score?: number | null
  focusedPct?: number | null
  generatedAt?: number | null
  // #47 D5 — AI data-quality counters (003 migration); null = counts unknown
  // (AI off, or a row written by an older build).
  confidentSamples?: number | null
  skippedSamples?: number | null
  // I83 — was AI focus detection on for this session? (004 migration.)
  // 1 = on, 0 = off, null = unknown. Every other AI column reads null both
  // when AI was off and when it was on but never produced a check; this is
  // what lets the report tell those two apart.
  aiEnabled?: number | null
  // Immutable local-owner provenance captured when the logical session first
  // starts. Historical reports must never substitute whichever identity is
  // active when they are opened.
  localEdPubkey?: string | null
  localDisplayName?: string | null
  // A same-topic stint whose in-memory score machine was not continuous (for
  // example after app restart) must replace, including clear, focus fields
  // rather than letting SQLite retain a misleading last-stint score.
  replaceFocusMetrics?: boolean
}

// Shape returned by `sessions_list` / `sessions_get`. Tauri auto-camelCases
// parameter names on JS→Rust invokes, but the response is serde's serialized
// struct, which uses Rust's snake_case field names verbatim.
export type SessionRecord = {
  id: string
  started_at: number | null
  ended_at: number | null
  total_minutes: number | null
  // Null/omitted means the row predates durable fractional-duration storage.
  total_duration_ms?: number | null
  peer_pubkeys: string | null
  // Optional at the TS boundary for mixed frontend/backend builds. Missing and
  // NULL both mean the row predates precise per-peer duration accounting.
  peer_presence_ms?: string | null
  declared_topic: string | null
  score: number | null
  focused_pct: number | null
  generated_at: number | null
  confident_samples: number | null
  skipped_samples: number | null
  ai_enabled: number | null
  // Optional at the TypeScript boundary for mixed frontend/backend builds;
  // missing has the same unknown-owner meaning as NULL from the migration.
  local_ed_pubkey?: string | null
  local_display_name?: string | null
}

export async function sessionsInsert(row: SessionRow): Promise<void> {
  await invoke('sessions_insert', {
    id: row.id,
    startedAt: row.startedAt,
    endedAt: row.endedAt,
    totalMinutes: row.totalMinutes,
    totalDurationMs: row.totalDurationMs,
    peerPubkeys: row.peerPubkeys,
    peerPresenceMs: row.peerPresenceMs ?? null,
    declaredTopic: row.declaredTopic ?? null,
    score: row.score ?? null,
    focusedPct: row.focusedPct ?? null,
    generatedAt: row.generatedAt ?? null,
    confidentSamples: row.confidentSamples ?? null,
    skippedSamples: row.skippedSamples ?? null,
    aiEnabled: row.aiEnabled ?? null,
    localEdPubkey: row.localEdPubkey ?? null,
    localDisplayName: row.localDisplayName ?? null,
    replaceFocusMetrics: row.replaceFocusMetrics ?? false,
  })
}

// Recovery path for an indeterminate sessions_get result. The Rust command's
// INSERT ... ON CONFLICT DO NOTHING is atomic: a first-ever stint is retained,
// while an existing multi-stint row can never be rewound by stint-only data.
export async function sessionsInsertIfAbsent(
  row: SessionRow
): Promise<boolean> {
  return invoke<boolean>('sessions_insert_if_absent', {
    id: row.id,
    startedAt: row.startedAt,
    endedAt: row.endedAt,
    totalMinutes: row.totalMinutes,
    totalDurationMs: row.totalDurationMs,
    peerPubkeys: row.peerPubkeys,
    peerPresenceMs: row.peerPresenceMs ?? null,
    declaredTopic: row.declaredTopic ?? null,
    score: row.score ?? null,
    focusedPct: row.focusedPct ?? null,
    generatedAt: row.generatedAt ?? null,
    confidentSamples: row.confidentSamples ?? null,
    skippedSamples: row.skippedSamples ?? null,
    aiEnabled: row.aiEnabled ?? null,
    localEdPubkey: row.localEdPubkey ?? null,
    localDisplayName: row.localDisplayName ?? null,
  })
}

export async function listSessions(): Promise<SessionRecord[]> {
  return invoke<SessionRecord[]>('sessions_list')
}

export async function sessionsGet(id: string): Promise<SessionRecord | null> {
  return invoke<SessionRecord | null>('sessions_get', { id })
}

// R4 — deletes the session row + its audit_events in one Rust transaction.
// `id` is the session topic.
export async function sessionsDelete(id: string): Promise<void> {
  await invoke('sessions_delete', { id })
}

// R4 — clears every session row and all audit_events in one Rust transaction.
export async function sessionsClearAll(): Promise<void> {
  await invoke('sessions_clear_all')
}

File: src-tauri/src/commands/sessions.rs
//! Session + audit-event commands: thin sync wrappers over `db::sessions` and
//! `db::audit_events` (same held-guard, no-async locking pattern as
//! `commands/friends.rs`). Upsert/cascade semantics live in the `db` layer;
//! the TS callers are `src/lib/db/sessions.ts` and `src/lib/db/audit.ts`
//! (camelCase invoke args in, snake_case serde rows out).

use tauri::{AppHandle, Runtime, State};

use crate::commands::session_journal;
use crate::db::{audit_events, session_timelines, sessions, DbPool};

fn lock<'a>(
    state: &'a State<'_, DbPool>,
) -> Result<std::sync::MutexGuard<'a, rusqlite::Connection>, String> {
    state.0.lock().map_err(|e| format!("db poisoned: {e}"))
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub fn sessions_insert(
    state: State<'_, DbPool>,
    id: String,
    started_at: i64,
    ended_at: i64,
    total_minutes: Option<i64>,
    total_duration_ms: Option<i64>,
    peer_pubkeys: Option<String>,
    peer_presence_ms: Option<String>,
    declared_topic: Option<String>,
    score: Option<i64>,
    focused_pct: Option<f64>,
    generated_at: Option<i64>,
    confident_samples: Option<i64>,
    skipped_samples: Option<i64>,
    // I83 — 1 = AI focus detection was on for this session, 0 = off, None =
    // caller didn't say (older frontend). None coalesces, so an omitted value
    // never overwrites a recorded one.
    ai_enabled: Option<i64>,
    local_ed_pubkey: Option<String>,
    local_display_name: Option<String>,
    // A focus score is stateful. If a same-topic stint began without its
    // predecessor's in-memory state, the frontend passes true so an honest
    // null can clear the stale score instead of COALESCE retaining it.
    replace_focus_metrics: Option<bool>,
) -> Result<(), String> {
    let conn = lock(&state)?;
    let row = sessions::SessionRow {
        id,
        started_at: Some(started_at),
        ended_at: Some(ended_at),
        total_minutes,
        total_duration_ms,
        peer_pubkeys,
        declared_topic,
        score,
        focused_pct,
        generated_at,
        confident_samples,
        skipped_samples,
        ai_enabled,
        local_ed_pubkey,
        local_display_name,
        peer_presence_ms,
    };
    sessions::insert_with_focus_metrics_mode(&conn, &row, replace_focus_metrics.unwrap_or(false))
        .map_err(|e| e.to_string())
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub fn sessions_insert_if_absent(
    state: State<'_, DbPool>,
    id: String,
    started_at: i64,
    ended_at: i64,
    total_minutes: Option<i64>,
    total_duration_ms: Option<i64>,
    peer_pubkeys: Option<String>,
    peer_presence_ms: Option<String>,
    declared_topic: Option<String>,
    score: Option<i64>,
    focused_pct: Option<f64>,
    generated_at: Option<i64>,
    confident_samples: Option<i64>,
    skipped_samples: Option<i64>,
    ai_enabled: Option<i64>,
    local_ed_pubkey: Option<String>,
    local_display_name: Option<String>,
) -> Result<bool, String> {
    let conn = lock(&state)?;
    let row = sessions::SessionRow {
        id,
        started_at: Some(started_at),
        ended_at: Some(ended_at),
        total_minutes,
        total_duration_ms,
        peer_pubkeys,
        declared_topic,
        score,
        focused_pct,
        generated_at,
        confident_samples,
        skipped_samples,
        ai_enabled,
        local_ed_pubkey,
        local_display_name,
        peer_presence_ms,
    };
    sessions::insert_if_absent(&conn, &row).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn sessions_list(state: State<'_, DbPool>) -> Result<Vec<sessions::SessionRow>, String> {
    let conn = lock(&state)?;
    sessions::list(&conn).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn sessions_get(
    state: State<'_, DbPool>,
    id: String,
) -> Result<Option<sessions::SessionRow>, String> {
    let conn = lock(&state)?;
    sessions::get(&conn, &id).map_err(|e| e.to_string())
}

// R4 + #236 — the SQL cascade drops the session, its audit events, and its
// written timeline in one transaction; the raw observation journal is a file,
// so it is unlinked after that transaction commits. Deleting rows first means a
// failure can only ever leave an orphaned journal (which the next regeneration
// or clear-all removes), never a narrative whose evidence is already gone.
#[tauri::command]
pub fn sessions_delete<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, DbPool>,
    id: String,
) -> Result<(), String> {
    {
        let mut conn = lock(&state)?;
        sessions::delete(&mut conn, &id).map_err(|e| e.to_string())?;
    }
    session_journal::remove_journal(&app, &id);
    Ok(())
}

#[tauri::command]
pub fn sessions_clear_all<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, DbPool>,
) -> Result<(), String> {
    {
        let mut conn = lock(&state)?;
        sessions::clear_all(&mut conn).map_err(|e| e.to_string())?;
    }
    session_journal::clear_journals(&app);
    Ok(())
}

// #236 — the post-session written timeline. `entries` is the JSON array the
// frontend validated before persisting; this layer stores it verbatim, the same
// way audit-event detail is stored.
#[tauri::command]
pub fn session_timeline_get(
    state: State<'_, DbPool>,
    session_id: String,
) -> Result<Option<session_timelines::SessionTimelineRow>, String> {
    let conn = lock(&state)?;
    session_timelines::get(&conn, &session_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn session_timeline_save(
    state: State<'_, DbPool>,
    session_id: String,
    generated_at: i64,
    model_id: Option<String>,
    source: String,
    entries: String,
    truncated: bool,
) -> Result<(), String> {
    let conn = lock(&state)?;
    let row = session_timelines::SessionTimelineRow {
        session_id,
        generated_at,
        model_id,
        source,
        entries,
        truncated: i64::from(truncated),
    };
    session_timelines::upsert(&conn, &row).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn audit_event_insert(
    state: State<'_, DbPool>,
    session_id: String,
    ts: i64,
    who: String,
    kind: String,
    detail: String,
    sig: String,
) -> Result<(), String> {
    let conn = lock(&state)?;
    let row = audit_events::AuditEventRow {
        session_id,
        ts,
        who,
        kind,
        detail,
        sig,
    };
    audit_events::insert(&conn, &row).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn audit_events_list_for_session(
    state: State<'_, DbPool>,
    session_id: String,
) -> Result<Vec<audit_events::AuditEventRow>, String> {
    let conn = lock(&state)?;
    audit_events::list_for_session(&conn, &session_id).map_err(|e| e.to_string())
}

// R7 — cross-session audit events for the local focus-insights view. The
// frontend shapes them in the pure statsInsights seam; this command only
// reads.
#[tauri::command]
pub fn audit_events_list_all(
    state: State<'_, DbPool>,
) -> Result<Vec<audit_events::AuditEventRow>, String> {
    let conn = lock(&state)?;
    audit_events::list_ai_distractions_all(&conn).map_err(|e| e.to_string())
}
